include/oneapi/tbb/collaborative_call_once.h

```diff
@@ -0,0 +1,222 @@
+            my_func();
+            return true;
+        }
+
+    private:
+        Func& my_func;
+    };
+
+    std::atomic<std::int64_t> ref_count{0};
+    std::atomic<bool> m_initialized{false};
+
+    // Storage with task_arena and contexts must be initialized only by winner thread
+    // by calling make_runner() method
+    union {
+        storage_t m_storage;
+    };
+
+    template<typename Fn>
+    void isolated_execute(Fn f) {
+        call_once_delegate<Fn> delegate(f);
+
+        r1::isolate_within_arena(delegate, reinterpret_cast<std::intptr_t>(this));
+    }
+
+public:
+    once_runner() {}
+
+    ~once_runner() {
+        spin_wait_while(ref_count, [&](std::int64_t value) { return value > 0; }, std::memory_order_acquire);
+        if (m_initialized.load(std::memory_order_relaxed)) {
+            m_storage.~storage_t();
+        }
+    }
+
+    void make_runner() {
+        new(&m_storage) storage_t();
+        m_initialized.store(true, std::memory_order_release);
+    }
+
+    void wait_for_init() {
+        spin_wait_while_eq(m_initialized, false);
+    }
+
+    void increase_ref() { ref_count++; }
+
+    void decrease_ref() { ref_count--; }
+
+    template <typename F>
+    void run_once(F&& f) {
+        m_storage.my_arena.execute([&] {
+            isolated_execute([&] {
+                function_stack_task<F> t{ std::forward<F>(f), m_storage.my_wait_context };
+                m_storage.my_wait_context.reserve();
+
+                execute_and_wait(t, m_storage.my_context, m_storage.my_wait_context, m_storage.my_context);
+            });
+        });
+    }
+
+    void assist() noexcept {
+        m_storage.my_arena.execute([&] {
+            isolated_execute([&] {
+                // We do not want to get an exception from user functor on moonlighting threads.
+                // The exception is handled with the winner thread
+                task_group_context stub_context;
+                wait(m_storage.my_wait_context, stub_context);
+            });
+        });
+    }
+
+};
+
+class runner_lifetime_checker : no_copy {
+    once_runner& m_runner;
+public:
+    runner_lifetime_checker(once_runner& r) : m_runner(r) {
+        m_runner.increase_ref();
+    }
+    ~runner_lifetime_checker() {
+        m_runner.decrease_ref();
+    }
+};
+
+class collaborative_once_flag : no_copy {
+    enum state {uninitialized, done};
+    std::atomic<std::uintptr_t> my_state{ state::uninitialized };
+
+    template <typename Fn, typename... Args>
+    friend void collaborative_call_once(collaborative_once_flag& flag, Fn&& f, Args&&... args);
+
+    template <typename Fn>
+    void do_collaborative_call_once(Fn&& f) {
+        std::uintptr_t expected = my_state.load(std::memory_order_acquire);
+        once_runner local_runner;
+
+        while (expected != state::done) {
+            if (expected == state::uninitialized && my_state.compare_exchange_strong(expected, reinterpret_cast<std::uintptr_t>(&local_runner))) {
```

Comment:
```suggestion
            if (expected == state::uninitialized && my_state.compare_exchange_strong(expected, local_runner.to_bits())) {
```
Apply in other places, as well